### PR TITLE
add initial node functionality

### PR DESCRIPTION
**Pull request by wmatray**
- Add linear layers to graph
- Add adaptive pooling node 
- Add circular padding node
- Remove node
- Basic adaptation using adaptive pooling and circular padding

### CODE REVIEW

File: src/core.py [hunk 1]
```diff
@@ -4,114 +4,193 @@
...
-    if node.op == "call_module":
-        submodule_name = node.target
-        if hasattr(graph, submodule_name):
-            delattr(graph, submodule_name)
 
+    # Adapt connections between input node and its new users
+    # In this case, the new node is the input node, so new_node input is always correct already
+    graph = adapt_connections(graph, new_node=input_node, 
+                              new_node_shape=(output_left_shape[-1], output_left_shape[-1]), 
+                              input_shape=output_left_shape, output_shape=input_right_shape)
```
Comment (PatrickNercessian) [inline src/core.py:143]: Is the last index not the batch dimension?

Have had the same question elsewhere in this PR

File: src/core.py [hunk 2]
```diff
@@ -4,114 +4,193 @@
...
+        
+        # Get shape of the new node
+        new_node_shape = (target_size, target_size)  # For repeat, in/out are same
 
     graph.graph.lint()
     graph.recompile()
         
-    return graph, name
+    # Fix the connections using pre-computed shapes
+    adapt_connections(graph, new_node, new_node_shape, input_shape, output_shape)
```
Comment (PatrickNercessian) [inline src/core.py:103]: Don't we need the input_shape param of adapt_connections to be the input of whatever comes AFTER the new node?

Assuming the reference node comes before the new node

**Review by PatrickNercessian — CHANGES_REQUESTED** [review]
A bunch of these are probably just me misunderstanding. A few code quality change requests though

File: src/add.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        new_node = graph.graph.call_module(
+            module_name=name,
+            args=(reference_node,),
+            kwargs={},
+        )
+
+    reference_node.replace_all_uses_with(new_node)
+    new_node.args = (reference_node,)
+
+    return graph, new_node
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/add.py]: I did this in https://github.com/PatrickNercessian/evolving-dnn/pull/2/commits/ec2659e83f2608e09149c265c8768fa0194b4ca5 so that you don't have to deal with my nitpicking. Let me know if you disagree, I'm happy to revert if so

File: src/core.py [hunk 4]
```diff
@@ -4,114 +4,193 @@
...
-            module_name=name,
-            args=(node,),
-            kwargs={},
-        )
-        node.replace_all_uses_with(new_node)
-        new_node.args = (node,)
+    # Add a linear layer to the graph
+    if operation == 'linear':
+        # get the shape of the reference node from metadata
+        reference_node_shape = reference_node.meta['tensor_meta'].shape
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/core.py:67]: I did this in https://github.com/PatrickNercessian/evolving-dnn/pull/2/commits/ec2659e83f2608e09149c265c8768fa0194b4ca5 so that you don't have to deal with my nitpicking. Let me know if you disagree, I'm happy to revert if so

File: src/add.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/add.py]: The only functional change in there I think is that I just use `module.__class__.__name__` for the param to get_unique_name(), to allow the new function to be abstract

[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/add.py]: Made it even cleaner https://github.com/PatrickNercessian/evolving-dnn/pull/2/commits/a1e54da85690834618f87d2aaa7ac25f2c2beb10

File: src/core.py [hunk 5]
```diff
@@ -4,114 +4,193 @@
...
+        # Get the first node (should be placeholder/input)
+        placeholder = next(iter(graph.graph.nodes))
+        placeholder.meta['tensor_meta'] = {
+            'dtype': example_input.dtype,
+            'shape': input_shape,
+            'requires_grad': example_input.requires_grad
+        }
+        
+        # Run shape propagation
+        ShapeProp(graph).propagate(example_input)
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:42]: It stores the output shape which is used in a lot of places

File: src/core.py [hunk 6]
```diff
@@ -4,114 +4,193 @@
...
-            kwargs={},
-        )
-        node.replace_all_uses_with(new_node)
-        new_node.args = (node,)
+    # Add a linear layer to the graph
+    if operation == 'linear':
+        # get the shape of the reference node from metadata
+        reference_node_shape = reference_node.meta['tensor_meta'].shape
+        # add a linear layer to the graph
+        graph, new_node = add_linear(graph, reference_node, reference_node_shape[-1])
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: For the linear layer that's all that matters

File: src/core.py (hunk 1 above)
Comment (wmatray) [reply src/core.py:159]: First index is batch

File: src/core.py (hunk 2 above)
Comment (wmatray) [reply src/core.py:119]: The verbiage can be a little confusing, "output_shape" is the required input for the users of the new node

File: src/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:85]: Good catch

File: src/core.py [hunk 7]
```diff
@@ -4,114 +4,193 @@
...
+        # get the shape of the new node from the module
+        new_node_shape = (graph.get_submodule(new_node.target).in_features,
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: Not if the new node's shapes are randomly generated

File: src/utils.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
+import torch
+import torch.nn as nn
+import math
+import torch.nn.functional as F
+import torch.fx
+from torch.fx.passes.shape_prop import ShapeProp
+import random
+
+
+def find_required_shapes(graph, node):
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/utils.py:10]: good catch

File: src/add.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/add.py]: Great idea, just be careful making changes, moving the functions around broke adapt_connections()

Fixed it in latest commit

File: src/core.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py]: I see what you mean, by reorganizing we already have it

File: src/core.py [hunk 9]
```diff
@@ -1,118 +1,196 @@
...
 
+def get_graph(model: nn.Module, input_shape: tuple):
 
-def get_graph_module(model):
     """
     Takes a PyTorch model and returns its computation graph using torch.fx
     
     Args:
         model: A PyTorch model (nn.Module)
+        input_shape: tuple specifying input tensor shape (batch, seq_len, dim)
```
Comment (MortimerE) [inline src/core.py:19]: A comment about why we need a specific input shape and cannot use ShapeProp's fake tensors would be helpful here, or a readme desc

File: src/core.py [hunk 10]
```diff
@@ -1,118 +1,196 @@
...
-    # ShapeProp(graph).propagate()
+    graph = IndividualGraphModule(torch.fx.symbolic_trace(model))
+    
+    # Perform shape propagation if input_shape is provided
+    if input_shape is not None:
+        # Create example input
+        example_input = torch.randn(input_shape)
+        
+        # Get the first node (should be placeholder/input)
+        placeholder = next(iter(graph.graph.nodes))
```
Comment (MortimerE) [inline src/core.py:33]: similar to input shape, just for tracking information while writing the paper as this becomes larger I think it would be good to backlink the docs for how the placeholder works 
The reason I want to do the readmes is so we can forget little details like this because we already wrote down something like "the placeholder acts as a root node that represents the input tensor to the network at a stage before any layers have processed it" 

I think giving gpt/claude the otter transcripts from meetings where we talk through it, the code snippets, and maybe the docs page should get it to give us like a 95% solid readme for stuff like this and then writing the paper and appendix sections presenting the experimental methods/tools will be a cinch

File: src/core.py (hunk 2 above)
Comment (MortimerE) [reply src/core.py:100]: From what I remember: 
parent->ref
parent->new->ref 

What happens if adapt connections or lint fails here? Should we have some kind of error handling or at least log the issues?

File: src/core.py [hunk 11]
```diff
@@ -1,118 +1,196 @@
...
+        reference_node: The node to remove
     Returns:
         graph: The modified graph
-        name: The name of the added module
     """
+    input_node = reference_node.args[0]
+    
+    # Get shapes before removing node
+    output_left_shape = list(input_node.meta['tensor_meta'].shape)
+    input_right_shape = reference_node.meta['tensor_meta'].shape
```
Comment (MortimerE) [inline src/core.py:125]: suggest rename to input_child and output_parent

File: src/core.py [hunk 12]
```diff
@@ -1,118 +1,196 @@
...
-    placeholder = next(iter(graph.graph.nodes))
-    placeholder.meta['tensor_meta'] = {
-        'dtype': example_input.dtype,
-        'shape': input_shape,
-        'requires_grad': example_input.requires_grad
-    }
     
-    # Run shape propagation
+    # Get required shapes before making any modifications
+    input_shape, output_shape = find_required_shapes(reference_node)
```
Comment (MortimerE) [inline src/core.py:61]: suggest rename to input_shape_new, output_shape_new 
Just bc input shape means something slightly different in get_graph_module 
Maybe we rename it there actually, something like data_shape?

**Review by MortimerE — APPROVED** [review]
I don't think any changes are needed at the current time, I just posted a handful of questions and ideas to keep in mind for future iterations that I/we can come back to later

File: src/core.py [hunk 13]
```diff
@@ -1,118 +1,196 @@
...
-    Removes a node from the graph and cleans up its module if it was dynamically added
+    Adapts the connections to/from a node to ensure all connected nodes have compatible shapes.
     
     Args:
         graph: The FX graph
-        node: The node to remove
+        new_node: The node whose connections need adaptation
+        new_node_shape: The shape of the new node
+        input_shape: The shape of the input node (pre-computed)
+        output_shape: The shape of the output node (pre-computed)
```
Comment (MortimerE) [inline src/core.py:163]: like above, suggestion to rename input_shape and output_shape to more specific names

File: src/utils.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+def find_required_shapes(node):
+    """
+    Finds the required shapes for a given node in the graph.
+    
+    Args:
+        graph: The FX graph
+        node: The node to find the required shapes for
+    Returns:
+        input_shape: The required shape of the input nodes, can be a list of shapes if the node has multiple inputs
+        output_shape: The required shape of the output node
```
Comment (MortimerE) [inline src/utils.py:19]: Suggestion to rename input_shape and output_shape to more specific names
I know this is in a different file but I think it just makes it more clear to read, just my opinion

File: src/core.py (hunk 9 above)
Comment (wmatray) [reply src/core.py:19]: Shapeprop needs an input shape to do shape propagation

File: src/core.py (hunk 11 above)
Comment (wmatray) [reply src/core.py:125]: I'm down to make this switch but might get confusing when we start using parent/child for mutation populations and crossover

File: src/core.py [hunk 15]
```diff
@@ -1,118 +1,196 @@
...
+    
+    # Get shapes before removing node
+    output_left_shape = list(input_node.meta['tensor_meta'].shape)
+    input_right_shape = reference_node.meta['tensor_meta'].shape
+    
+    # Remove the node from the graph
+    reference_node.replace_all_uses_with(input_node)
+    
+    # Delete the submodule
+    delattr(graph, reference_node.target)
```
[the remark this answers is not in the document]
Comment (wmatray) [reply src/core.py:131]: That's a great point that I never considered.

At first I thought it can't be possible because our graph is topologically ordered; a node can never be an input to a node above it in the order. But what if I manually created a forward pass like this: x -> linear1() -> linear2() -> linear1(). The graph built by torch fx includes 3 nodes, considering the second linear1() call to be a distinct node that calls the same module.

Will fix
